### PR TITLE
[JENKINS-35696] - Allow registering BC on the slave

### PR DESCRIPTION
**Pull request by alvarolobato**
New method on the API to allow registering BC on the slave. The plugin is responsible for calling the method with a Channel to the slave.

Multiple calls to register BC on the same Channel will have no effect.

[JENKINS-35696](https://issues.jenkins-ci.org/browse/JENKINS-35696)

### CODE REVIEW

**Comment by alvarolobato** [comment]
@reviewbybees esp @amuniz @stephenc

File: src/main/java/jenkins/bouncycastle/api/PEMEncodable.java [hunk 1]
```diff
@@ -487,5 +491,25 @@ private static String encodeBase64(@Nonnull byte[] data) {
...
+     * Registers bouncy castle on the slave JVM
+     * 
+     * @param channel to the slave
+     * @return result of the operation
+     * @throws Exception if there is a problem registering bouncycastle
+     */
+    public static boolean registerBCOnSlave(@Nonnull Channel channel) throws Exception {
+        if (channel.getProperty(BOUNCYCASTLE_REGISTERED) == null) {
+            // pre-loading the bouncyclastle jar to make sure the JVM reconizes the signature
+            channel.preloadJar(PEMEncodable.class.getClassLoader(), BouncyCastleProvider.class);
```
Comment (amuniz) [inline src/main/java/jenkins/bouncycastle/api/PEMEncodable.java]: I've never used `preloadJar`. Is the plugin classloader the right one to use here?

**Comment by amuniz** [comment]
:bee:

File: src/main/java/jenkins/bouncycastle/api/PEMEncodable.java (hunk 1 above)
Comment (alvarolobato) [inline src/main/java/jenkins/bouncycastle/api/PEMEncodable.java]: I think so, AFIK you need a class loader that know where the correct class is loaded from, in this case BouncyCastleProvider.class

**Comment by andresrc** [comment]
I'd include some synchronization on the channel (indirectly, like the guava cache using weak references in the keys) in case more than one plugin tries to register in parallel.

Apart from that, if the class loader used is the correct one :bee:

**Comment by alvarolobato** [comment]
@andresrc yes it can be done, but the addProvider is an innocuous operation, the second time does nothing. And the preloadJar is just a preload... so maybe it's better to keep it simple. Thanks for the review

**Comment by stephenc** [comment]
The local-side property value can be a Future for completion. That way if it is missing you have an almost idempotent write and worst case make two remoting calls before completing the future. All subsequent code that sees the future can just get the future ... **and for the win** we can rethrow the fiture's exception if we failed so that the failure reason is available to all, not burried in the logs that were rotated away because the system has been up too long

File: src/main/java/jenkins/bouncycastle/api/BCRegisterer.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import javax.annotation.Nonnull;
+
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+import jenkins.security.MasterToSlaveCallable;
+
+/**
+ * Callable to start the remote agent.
+ */
+public class BCRegisterer extends MasterToSlaveCallable<Boolean, Exception> {
```
Comment (stephenc) [inline src/main/java/jenkins/bouncycastle/api/BCRegisterer.java]: @Restricted as you have a utility method to manage the call

File: src/main/java/jenkins/bouncycastle/api/PEMEncodable.java [hunk 3]
```diff
@@ -487,5 +491,25 @@ private static String encodeBase64(@Nonnull byte[] data) {
...
     }
 
+    /**
+     * Registers bouncy castle on the slave JVM
+     * 
+     * @param channel to the slave
+     * @return result of the operation
+     * @throws Exception if there is a problem registering bouncycastle
+     */
+    public static boolean registerBCOnSlave(@Nonnull Channel channel) throws Exception {
```
Comment (stephenc) [inline src/main/java/jenkins/bouncycastle/api/PEMEncodable.java]: Hmmm I though you were putting this in a sensible utility class. What has this to do with registering BC.

Perhaps the callable should not be @Restricted, rather have a private constructor and put the utility method on the class directly

**Comment by stephenc** [comment]
:bug: the static utility method is not in an intuitive class

**Comment by stephenc** [comment]
Also see my **for the win** comment wrt exceptions

**Comment by alvarolobato** [comment]
@stephenc I've moved the method to the Callable and made the constructor private. I'm returning the Future from the remote call.

I'll have to leave the Exception re-throwing for a future improvement as I don't have time right now and we need to cut a version. I've created [JENKINS-35914](https://issues.jenkins-ci.org/browse/JENKINS-35914). I will involve creating the future (not returning the one from the asyncCall) and execute the preloadJar and the remote call all together.

**Comment by stephenc** [comment]
@alvarolobato https://github.com/alvarolobato/bouncycastle-api-plugin/pull/1 is the improvements that should be made IMHO. I hope you can see that this is not as hard... basically you get the re-throw for free and using ChannelProperty means that we:

1. Know it won't be serialized
2. Know nobody else can peek into the property

**Comment by alvarolobato** [comment]
Thanks @stephenc,  looks great. I don't have that knowledge about the API (yet), commiting

**Comment by stephenc** [comment]
:bee:

**Comment by andresrc** [comment]
:bee:

**Comment by amuniz** [comment]
:bee:

**Pull request merged**